### PR TITLE
Spectral regrouping by HEALPix pixel

### PR DESCRIPTION
**Pull request by tskisner**
This work is focused on grouping spectra by healpix pixel and performing that re-grouping on either a full production or on an incremental update from a few frames:

 - A new Spectra class can hold spectral data (wavelength, flux, ivar,
   mask, resolution, etc) for one or more arbitrary bands.  There are
   methods for returning a selection of spectra and for updating or
   appending new spectral data.  Each band may have different number
   of diagonals in the resolution matrix.

 - The pipeline is now designed to work with these spectra objects rather
   than bricks.

 - A new MPI-aware entry point (desispec.scripts.group_spectra) is added
   which distributes spectral group files among processes to perform the
   regrouping.  Each process reads (and optionally caches) the frames
   needed for its spectral groups.

 - Pipeline production creation now builds nightly chains of jobs which
   can run in parallel.

 - Redshift fitting will be broken until Redrock can be fully integrated
   and using the new spectral group files.

### CODE REVIEW

**Comment by sbailey** [comment]
Regarding failing Travis tests:

@weaver do you know why travis tests aren't finding healpy even though it is included in the PIP_DEPENDENCIES?

@tskisner have you tried putting healpy into CONDA_DEPENDENCIES or CONDA_ALL_DEPENCENCIES instead?

**Comment by moustakas** [comment]
I'm not at my computer but check where I added the dependency in the
sky-and-QA branch of desitarget, which worked for me.

File: py/desispec/io/spectra.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    hdus = fits.open(infile, mode="readonly")
+    nhdu = len(hdus)
+
+    # load the metadata.  FITS will capitalize all our metadata
+    # which is annoying.  Here we lower it all.
+
+    meta = {}
+    for k, v in hdus[0].header.items():
+        if k != "EXTNAME":
+            meta[k.lower()] = v
```
Comment (sbailey) [inline py/desispec/io/spectra.py]: Elsewhere we leave these as upper case.  Let's just cast this to a dictionary via `meta = dict(hdus[0].header)` and leave them as is.  Or you could even leave it as a `astropy.io.fits.header.Header` object in which case either upper or lowercase keys will work.  Header object are much less efficient to access, but that only matters if we are looking up keywords many many times.

**Comment by weaverba137** [comment]
@sbailey, it might help if you point out a specific test.

**Comment by weaverba137** [comment]
PS, I was going to say, first of all, that healpy should be a conda package.

**Comment by weaverba137** [comment]
* For the sphinx problem, you need to add healpy to `autodoc_mock_imports` in doc/conf.py (see line 136).
* For the test failures, there are additional failures that appear to be unrelated to healpy, so you'll need to fix those as well.

**Comment by sbailey** [comment]
sphinx fixed and tests can see healpy now, but now we have our favorite astropy making random changes on us type of error:
```
======================================================================
ERROR: test_empty (desispec.test.test_spectra.TestSpectra)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/desihub/desispec/py/desispec/test/test_spectra.py", line 137, in test_empty
    path = write_spectra(self.filebuild, spec)
  File "/home/travis/build/desihub/desispec/py/desispec/io/spectra.py", line 120, in write_spectra
    all_hdus.writeto("{}.tmp".format(outfile), overwrite=True, checksum=True)
TypeError: writeto() got an unexpected keyword argument 'overwrite'
```
overwrite is astropy 1.3, clobber is astropy 1.2, we should support both until 1.4 is out. :(  Other codes have handled this with try/except blocks on the writing.

**Comment by tskisner** [comment]
Hi folks, I don't believe healpy is available as a conda package, which is why I included it in pip dependencies.  I'll go add boilerplate for astropy.  Super annoying...

**Comment by weaverba137** [comment]
@tskisner, just so you know, the tests that required healpy did pass when it was installed with conda, so I guess that means there's a package.

**Comment by tskisner** [comment]
Your travis builds must have access to other conda channels then.  It is not in the default one:
```
$ which conda
/home/kisner/software/intel/idp/bin/conda
```
```
$ conda --version
conda 4.3.21
```
```
$ conda config --get channels
--add channels 'defaults'   # lowest priority
--add channels 'file:///home/kisner/software/intel/idp/conda-bld'   # highest priority
```
```
$ conda search healpy
Fetching package metadata ...........
$
```

A web search seems to show one version available on conda-forge though.

**Comment by tskisner** [comment]
I'll merge this unless there are further changes you would like...

**Pull request merged**